api/types/registry: move `ServiceConfig` legacy field marshaling support into daemon backend

austinvazquez opened the pull request:
**- What I did**
Partial of #50740

This change removes the `ServiceConfig.MarshalJSON` function from and moves the marshaling logic of legacy fields down into the daemon backend.

**- How I did it**

**- How to verify it**

**- Human readable description for the release notes**
```markdown changelog
api/types/registry: moved `ServiceConfig` legacy field marshaling support into daemon backend
```

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented on api/types/registry/registry.go:
```diff
@@ -17,23 +17,6 @@ type ServiceConfig struct {
 	ExtraFields map[string]any `json:"-"`
 }
 
-// MarshalJSON implements a custom marshaler to include legacy fields
-// in API responses.
-func (sc *ServiceConfig) MarshalJSON() ([]byte, error) {
```
It's used by the daemon to produce API responses for older API versions. If we remove this here, we probably need to create a wrapper struct with marshaling in the daemon API server; see https://github.com/moby/moby/blob/ed8a6a89cad73af1e54eb4e94bdf7c0f46a5737b/daemon/server/router/system/system_routes.go#L105-L112

- https://github.com/moby/moby/pull/49749

thaJeztah replied:
I recall I had a branch somewhere with a more generic package to allow "append old API fields" but not sure if I have preserved it 🙈

thaJeztah commented:
> Need to verify again if it is used by any of the popular Docker clients.

It should not be used by clients, only the daemon, but yeah, we need to find a good spot / approach for these. The general attempt here was to allow the daemon to continue serving "legacy" responses (where no sensible content is to be served, other than "the field is there") but being able to remove the fields from the structs before the API is a module (so basically no more opportunities to remove such fields if they're used by the client).

austinvazquez commented:
Okay that makes more sense. Thanks @thaJeztah. Let me rework this to create a structure down in the daemon and have the legacy fields support there.

austinvazquez committed:
daemon: minor touchups for backend system info response marshal

Signed-off-by: Austin Vazquez <austin.vazquez@docker.com>

austinvazquez committed:
daemon: marshal legacy `registry.ServiceConfig` extra fields for compatability

This change moves the logic that is used to marshal the legacy extra fields for `registry.ServiceConfig` type to the daemon backend.

Signed-off-by: Austin Vazquez <austin.vazquez@docker.com>

thaJeztah commented on daemon/server/router/system/info_response_test.go:
```diff
@@ -31,3 +34,58 @@ func TestLegacyFields(t *testing.T) {
+		},
+	}
+
+	for _, tt := range tests {
```
No need to change; we started to converge on `tc` ("test case") for these. Definitely no need to change, but it helps having some consistency and reduce cognitive load when reading ☺️

thaJeztah commented on daemon/server/router/system/info_response_test.go:
```diff
@@ -31,3 +34,58 @@ func TestLegacyFields(t *testing.T) {
+			assert: func(t *testing.T, data []byte, err error) {
+				assert.NilError(t, err)
+				for k, v := range expected {
+					assert.Check(t, strings.Contains(string(data), v), fmt.Sprintf("Expected legacy field %q to be included in the output: %s", k, string(data)))
```
I think `assert.Check(t, is.Contains(string(data), v))` may also work for strings, and print the output when failing

(writing from my phone so can't verify what it looks like)

corhere commented on daemon/server/router/system/info_response_test.go:
```diff
@@ -31,3 +34,58 @@ func TestLegacyFields(t *testing.T) {
+			assert: func(t *testing.T, data []byte, err error) {
+				assert.NilError(t, err)
+				for _, v := range expected {
+					assert.Check(t, !is.Contains(string(data), v)().Success())
```
String comparisons? As an alternative, you could [unmarshal into a `map[string]any`](https://go.dev/play/p/hVQdLtLwp27) and assert on the presence/absence of map keys.

austinvazquez replied:
Yep that looks much better. :)
